### PR TITLE
Feature/dhscft 1080 modular spine chart

### PR DESCRIPTION
**Pull request by stuc-bjss**
# Description

Jira ticket: [DHSCFT-1080](https://bjss-enterprise.atlassian.net/browse/DHSCFT-1080)

Make spine chart more modular

## Changes

- Added quartile hooks, but seeded quartile data on views to reduce complexity on chart main page
- Added hook to deal with multiple indicatorMeta data
- Added hook to collect from health data from multiple indicators
- Relocated SpineChart and associated components into charts/SpineChart

## Validation

Kept as many existing tests as possible
No changes at all to e2e tests

### CODE REVIEW

File: frontend/fingertips-frontend/components/charts/SpineChart/SpineChartTable/SpineChartTable.tsx [hunk 1]
```diff
@@ -47,6 +47,8 @@ export function SpineChartTable({
     return convertSpineChartTableToCsv(sortedData);
   }, [sortedData]);
 
+  if (!indicatorData.length) return null;
```
Comment (williamwillis11) [inline frontend/fingertips-frontend/components/charts/SpineChart/SpineChartTable/SpineChartTable.tsx:50]: worth adding a test for this early return into

frontend/fingertips-frontend/components/charts/SpineChart/SpineChartTable/SpineChartTable.test.tsx

File: frontend/fingertips-frontend/components/charts/SpineChart/SpineChartWrapper.tsx [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+export function SpineChartWrapper() {
+  const searchState = useSearchStateParams();
+  const { [SearchParams.BenchmarkAreaSelected]: benchmarkAreaSelected } =
+    searchState;
+
+  const benchmarkToUse = determineBenchmarkToUse(benchmarkAreaSelected);
+
+  const spineChartIndicatorData = useSpineChartData();
+  if (!spineChartIndicatorData) return null;
```
Comment (williamwillis11) [inline frontend/fingertips-frontend/components/charts/SpineChart/SpineChartWrapper.tsx:18]: worth adding a test for this return null into?:

frontend/fingertips-frontend/components/charts/SpineChart/SpineChartWrapper.test.tsx

File: frontend/fingertips-frontend/components/charts/SpineChart/helpers/convertSpineChartTableToCsv.ts [hunk 3]
```diff
@@ -64,7 +64,10 @@ export const convertSpineChartTableToCsv = (
 
     if (groupData) {
       const { areaCode, areaName, healthData } = groupData;
+      if (!healthData[0]) return;
```
Comment (williamwillis11) [inline frontend/fingertips-frontend/components/charts/SpineChart/helpers/convertSpineChartTableToCsv.ts:67]: worth a test?

File: frontend/fingertips-frontend/components/charts/SpineChart/helpers/convertSpineChartTableToCsv.ts [hunk 4]
```diff
@@ -86,7 +89,10 @@ export const convertSpineChartTableToCsv = (
 
     if (englandData && englandData.areaCode !== groupData?.areaCode) {
       const { areaCode, areaName, healthData } = englandData;
+      if (!healthData[0]) return;
```
Comment (williamwillis11) [inline frontend/fingertips-frontend/components/charts/SpineChart/helpers/convertSpineChartTableToCsv.ts:92]: worth a test?

File: frontend/fingertips-frontend/components/charts/SpineChart/helpers/spineChartIsRequired.test.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+      [SearchParams.AreasSelected]: ['A1', 'A2', 'A3'],
+      [SearchParams.IndicatorsSelected]: ['100', '101'],
+      [SearchParams.GroupAreaSelected]: 'GROUP_X',
+    };
+
+    const result = spineChartIsRequired(searchState);
+    expect(result).toBe(false);
+  });
+
+  it('returns false when indicatorsSelected.length < 2', () => {
```
Comment (williamwillis11) [inline frontend/fingertips-frontend/components/charts/SpineChart/helpers/spineChartIsRequired.test.ts:28]: can we have a case on the boundary indicatorsSelected.length = 2?

File: frontend/fingertips-frontend/components/charts/SpineChart/helpers/spineChartIsRequired.test.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+      [SearchParams.AreasSelected]: ['A1', 'A2'],
+      [SearchParams.IndicatorsSelected]: ['100', '101'],
+      [SearchParams.GroupAreaSelected]: ALL_AREAS_SELECTED,
+    };
+
+    const result = spineChartIsRequired(searchState);
+    expect(result).toBe(false);
+  });
+
+  it('returns false when all conditions fail', () => {
```
Comment (williamwillis11) [inline frontend/fingertips-frontend/components/charts/SpineChart/helpers/spineChartIsRequired.test.ts:50]: is it worth a test when indicatorsSelected is empty/0

and one when AreasSelected is  empty/0

?

File: frontend/fingertips-frontend/components/charts/SpineChart/helpers/spineChartIsRequired.test.ts (hunk 5 above)
Comment (williamwillis11) [reply frontend/fingertips-frontend/components/charts/SpineChart/helpers/spineChartIsRequired.test.ts:28]: maybe combined with the comment above so both on the boundary?

File: frontend/fingertips-frontend/components/molecules/TimePeriodDropDown/timePeriodDropDown.test.tsx [hunk 7]
```diff
@@ -1,5 +1,5 @@
 // MUST BE AT THE TOP
```
Comment (williamwillis11) [inline frontend/fingertips-frontend/components/molecules/TimePeriodDropDown/timePeriodDropDown.test.tsx:1]: // MUST BE AT THE TOP DUE TO HOISTING OF MOCKED MODULES

File: frontend/fingertips-frontend/components/charts/SpineChart/hooks/useSpineChartData.test.ts [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    ] = mockBuildSpineChartIndicatorData.mock.calls[0];
+
+    expect(Array.isArray(combinedHealthData)).toBe(true);
+    expect(indicatorMetaData).toEqual([indicator1, indicator2]);
+    expect(quartileData).toEqual([mockQuartileData()]);
+    expect(areaCodes).toEqual(['E09000002', 'E09000003']);
+    expect(selectedGroupCode).toEqual('E12000007');
+  });
+
+  it('returns undefined if insufficient data is available', async () => {
```
Comment (williamwillis11) [inline frontend/fingertips-frontend/components/charts/SpineChart/hooks/useSpineChartData.test.ts:83]: dont think we need async on this test? might be wrong

File: frontend/fingertips-frontend/components/charts/SpineChart/helpers/healthDataRequestAreas.test.ts [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    const searchState = {
+      [SearchParams.AreasSelected]: [areaCodeForEngland],
+      [SearchParams.AreaTypeSelected]: 'LocalAuthority',
+    };
+
+    const result = healthDataRequestAreas(searchState, [mockArea1]);
+
+    expect(result).toEqual([
+      {
+        areaCodes: [areaCodeForEngland],
```
Comment (alibombaye) [inline frontend/fingertips-frontend/components/charts/SpineChart/helpers/healthDataRequestAreas.test.ts:20]: would it make sense as part of this test to include more selected areas since we are checking only the selectedAreas are returned if england is included.

File: frontend/fingertips-frontend/components/charts/SpineChart/helpers/spineChartRequestParams.test.ts [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+        latestOnly: true,
+        benchmarkRefType: 'England',
+      });
+
+      expect(result).toContainEqual({
+        indicatorId: Number(indicators[1]),
+        areaCodes: [group],
+        areaType: groupType,
+        latestOnly: true,
+        benchmarkRefType: 'England',
```
Comment (alibombaye) [inline frontend/fingertips-frontend/components/charts/SpineChart/helpers/spineChartRequestParams.test.ts:79]: Use the const value in the types `BenchmarkReferenceType.England` instead for all these references to benchmarkRefType

File: frontend/fingertips-frontend/components/charts/SpineChart/helpers/spineChartIsRequired.test.ts [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+      [SearchParams.AreasSelected]: ['A1', 'A2'],
+      [SearchParams.IndicatorsSelected]: ['100', '101'],
+      [SearchParams.GroupAreaSelected]: 'GROUP_X',
+    };
+
+    const result = spineChartIsRequired(searchState);
+    expect(result).toBe(true);
+  });
+
+  it('returns false when areaCodes.length > 2', () => {
```
[the remark this answers is not in the document]
Comment (stuc-bjss) [reply frontend/fingertips-frontend/components/charts/SpineChart/helpers/spineChartIsRequired.test.ts:17]: the test before is exactly the same with 2 areas

**Review by williamwillis11 — APPROVED** [review]
